Initial job docs for PATh (#10)
osg-htc/user-documentation: documentation/PATh/htc_workloads/submitting_workloads/quickstart.md

---
path:
    path: htc_workloads/submitting_workloads/quickstart.md
---

# Quickstart - Submit Example HTCondor Jobs

## Job 1: A simple, nonparallel job

> First create a `short.sh` file to use as en executable for our sample job:

Reviewer: ```suggestion
First, using a text editing program (e.g. Nano, Vim), create a file called `short.sh` to use as an executable for our sample job:
```

	#!/bin/bash
	# short.sh: a short discovery job
    set -e
	printf "Start time: "; /bin/date
	printf "Job is running on node: "; /bin/hostname
	printf "Job running as user: "; /usr/bin/id
	printf "Job is running in directory: "; /bin/pwd
	echo
	echo "Working hard..."
	sleep 20
	echo "Science complete!"

Now, make the script executable.

> 	chmod +x short.sh

Reviewer 2: ```suggestion
	$ chmod +x short.sh
```

### Run the job locally

> When setting up a new job submission, it's important to test your job outside
> of HTCondor before submitting into pool. 

Reviewer: ```suggestion
When possible, it is important to first test your job on your local resorces prior to submitting the job to HTCondor to run on the PATh Facility execution points.  This will help identify potential errors within the code prior to queuing many HTCondor jobs. 

For example, we can run this executable on our local terminal by typing: 
```

	$ ./short.sh
	Start time: Wed Aug 21 09:21:35 CDT 2013
	Job is running on node: ap1.path-facility

> 	Job running as user: uid=54161(netid) gid=1000(users)

Reviewer: ```suggestion
	Job running as user: uid=54161(username) gid=1000(users)
```

> 	Job is running in directory: /home/netid/quickstart

Reviewer: ```suggestion
	Job is running in directory: /home/username/quickstart
```

	Working hard...
	Science complete!

### Create an HTCondor submit file

> So far, so good! Let's create a simple (if verbose) HTCondor submit file. Name it `tutorial01.submit`.

Reviewer: ```suggestion
So far, so good! Let's create a simple (if verbose) HTCondor submit file by using a text editing program. Name it `tutorial01.submit`.
```

	#  information about job execution steps.
	error = short.error
	output = short.output
	log = short.log

	# We need to request the resources that this job will need:
	request_cpus = 1
	request_memory = 100 MB
	request_disk = 1 GB

	# The last line of a submit file indicates how many jobs of the above
	#  description should be queued. We'll start with one job.
	queue 1

### More about projects

The PATh Facility is using projects to track usage and charge the
correct allocation. If you only have one project, you do not need to 
worry about this - the system will map your jobs automatically to 
that one project. However, if you have multiple ones, you will
have to add an attribute to your submit file to indicate which 
project to charge. The attribtute name is `accounting_group`.
For example, if your project is `AmazingScience`, the line in
the submit file (somewhere before the `queue` line) should
be:

    accounting_group = AmazingScience

### Submit the job 

Submit the job using `condor_submit`:

	$ condor_submit tutorial01.submit
	Submitting job(s). 
	1 job(s) submitted to cluster 144121.

### Check the job status

> The `condor_q` command tells the status of currently running jobs.

Reviewer: ```suggestion
The `condor_q` command tells the status of your jobs currently in the queue.
```

> Generally you will want to limit it to your own jobs: 

Reviewer: ```suggestion

```

	$ condor_q netid
	OWNER	   BATCH_NAME     SUBMITTED   DONE   RUN    IDLE  TOTAL JOB_IDS

> 	netid	 ID: 1441271  12/10 14:18	 _	1      _      1 1441271.0

Reviewer: ```suggestion
	username	 ID: 1441271  12/10 14:18	 _	1      _      1 1441271.0
```

	Total for query: 1 jobs; 0 completed, 0 removed, 0 idle, 1 running, 0 held, 0 suspended

> 	Total for netid: 1 jobs; 0 completed, 0 removed, 0 idle, 1 running, 0 held, 0 suspended

Reviewer: ```suggestion
	Total for username: 1 jobs; 0 completed, 0 removed, 0 idle, 1 running, 0 held, 0 suspended
```

	Total for all users: 3001 jobs; 0 completed, 0 removed, 2189 idle, 754 running, 58 held, 0 suspended

You can also get status on a specific job cluster: 

	$ condor_q 1441271
	OWNER	   BATCH_NAME     SUBMITTED   DONE   RUN    IDLE  TOTAL JOB_IDS

> 	netid	 ID: 1441271  12/10 14:18	 _	1      _      1 1441271.0

Reviewer: ```suggestion
	username	 ID: 1441271  12/10 14:18	 _	1      _      1 1441271.0
```

	Total for query: 1 jobs; 0 completed, 0 removed, 0 idle, 1 running, 0 held, 0 suspended

> 	Total for netid: 1 jobs; 0 completed, 0 removed, 0 idle, 1 running, 0 held, 0 suspended

Reviewer: ```suggestion
	Total for username: 1 jobs; 0 completed, 0 removed, 0 idle, 1 running, 0 held, 0 suspended
```

	Total for all users: 3001 jobs; 0 completed, 0 removed, 2189 idle, 754 running, 58 held, 0 suspended
	

> Note the `DONE`, `RUN`, and `IDLE` columns. Your job will be listed in the `IDLE` column if

Reviewer: ```suggestion
Note the `DONE`, `RUN`, and `IDLE` columns. Your job will be listed in the `IDLE` column if it is waiting to match to an execute machine and 
```

it hasn't started yet. If it's currently scheduled and running, it will
appear in the `RUN` column. As it finishes up, it will then show in the `DONE` column.
Once the job completes completely, it will not appear in `condor_q`. 

Let's wait for your job to finish – that is, for `condor_q` not to show
the job in its output. A useful tool for this is watch – it runs a
program repeatedly, letting you see how the output differs at fixed
time intervals. Let's submit the job again, and watch `condor_q` output
at two-second intervals: 

	$ condor_submit tutorial01.submit
	Submitting job(s). 
	1 job(s) submitted to cluster 1441272

> 	$ watch -n2 condor_q netid 

Reviewer 2: ```suggestion
	$ condor_watch_q
```

	... 

When your job has completed, it will disappear from the list. 

*Note*: To close watch, hold down Ctrl and press C. 

### Job history

Once your job has finished, you can get information about its execution
from the `condor_history` command: 

	$ condor_history 1441272
	 ID      OWNER            SUBMITTED     RUN_TIME ST   COMPLETED CMD           
	 1441272.0   netid     12/10 14:18   0+00:00:29 C  12/10 14:19 /home/netid/tutorial-quickstart/short.sh 

*Note*: You can see much more information about your job's final status
using the `-long` option. 

### Check the job output

Once your job has finished, you can look at the files that HTCondor has
returned to the working directory. The names of these files were specified in our
submit file. If everything was successful, it should have returned:

* a log file from HTCondor for the job cluster: short.log
* an output file for each job's output: short.output

> * an error file for each job's errors: short.error

Reviewer: ```suggestion
* an error file for each job's standard errors: short.error
```

Read the output file. It should be something like this: 

	$ cat short.output
	Start time: Mon Dec 10 20:18:56 UTC 2018
	Job is running on node: osg-84086-0-cmswn2030.fnal.gov
	Job running as user: uid=12740(osg) gid=9652(osg) groups=9652(osg)
	Job is running in directory: /srv

	Working hard...
	Science complete!
	
Job 2: Passing arguments to executables 
---------------------------------------

Sometimes it's useful to pass arguments to your executable from your
submit file. For example, you might want to use the same job script
for more than one run, varying only the parameters. You can do that
by adding Arguments to your submission file.

First, let's edit our existing `short.sh` script to accept arguments. To
avoid losing our original script, we make a copy of the file under the
name `short_transfer.sh`

	$ cp short.sh short_transfer.sh
	
Now, edit the file to include the added lines below:

	#!/bin/bash
	# short.sh: a short discovery job

>     set -e

Reviewer 2: ```suggestion
        set -e
```

    job.1444786.2.log     job.1444786.3.log	    job.1444786.4.log
    job.1444786.5.log     job.1444786.6.log     job.1444786.7.log
    job.1444786.8.log     job.1444786.9.log     job.1444786.0.output
    job.1444786.1.output  job.1444786.2.output  job.1444786.3.output
    job.1444786.4.output  job.1444786.5.output  job.1444786.6.output
    job.1444786.7.output  job.1444786.8.output  job.1444786.9.output

### Where did jobs run? 

It might be interesting to see where our jobs actually ran. To get that
information for a single job, we can use the command `condor_history`.
First, select a job you want to investigate and then run `condor_history
-long jobid`. In this example, we will investigate the job ID 1444786.9.
Again the output is quite long:

	$ condor_history -limit 1 -long 1444786.9
	BlockWriteKbytes = 0
	BlockReads = 0
	DiskUsage_RAW = 36
	... 
	MATCH_EXP_JOBGLIDEIN_ResourceName = "Georgia_Tech_PACE_CE_2"
	...
	
Looking through here for a hostname, we can see that the parameter
`MATCH_EXP_JOBGLIDEIN_ResourceName`. The output of this parameter is
the slot our job ran on. In this example, our job ran on a slot at the

> Georgia Institute of Technology.

Reviewer: The path facility isn't connected to the OSPool
